### DISCUSSION TITLE
[Compilers] Compiler generates different bytecode for unmodified proxy contract

### DISCUSSION THREAD

**Discussion by uF4No** [zkSync-Community-Hub/zksync-developers#703] [Compilers] [answered]
### Which compiler are you using?

- [X] zksolc
- [ ] zkvyper

### Which version of Solidity or Vyper?

0.8.19

### zksolc/zkvyper version?

1.5.2

### Explain the issue

[Original question from @Haypierre in this discussion](https://github.com/zkSync-Community-Hub/zksync-developers/discussions/639#discussioncomment-10693912).

> I'm now working on batch transactions support for openfort smart accounts on zkSync.
> I have a [dedicated branch](https://github.com/openfort-xyz/openfort-zksync-contracts/tree/batch_calls) with the in-progress implementation.
> When I edit a single byte of BaseOpenfortAccount.sol, hardhat-zksync produces a different bytecode for the UpgradeableOpenfortProxy when It should not.
> 
> UpgradeableOpenfortProxy never changes. Therefore, its bytecode and bytecode hash should remain constant.
Deploying our Factory with a wrong proxy bytecode hash causes the smart accounts to have no associated bytecodes after being deployed from the createAccountWithNonce factory function.
> 
> I'm sure I'm doing something wrong.
>
> Here is a PR explaining the issue:
> https://github.com/openfort-xyz/openfort-zksync-contracts/pull/4/files
>
> When I hard code the UpgradeableOpenfortProxy bytecode and bytecode hash in the factory deployment hardhat task, it works fine:
> 
> era_test_node fork mainnet
> yarn hardhat test --grep "self-custody account flow: sign raw transaction" --nonce 13

### Link to your repo?

https://github.com/openfort-xyz/openfort-zksync-contracts/tree/batch_calls

**Comment by alhenaai** [comment]
It seems like you are encountering an issue where modifying a single byte in `BaseOpenfortAccount.sol` results in a different bytecode for `UpgradeableOpenfortProxy`, which should not be affected by these changes. This issue might be related to how the compiler handles dependencies or the generation of bytecode.

Here are a few steps you can take to troubleshoot and potentially resolve this issue:

1. **Check Compiler Settings**: Ensure that your compiler settings in `hardhat.config.ts` are consistent and not causing recompilation of unchanged contracts. Sometimes, global settings like optimization can affect the output unexpectedly.

2. **Dependency Check**: Verify if `BaseOpenfortAccount.sol` has any indirect influence or dependency on `UpgradeableOpenfortProxy`. Even if it seems unrelated, the compiler might be including some linked information that causes the bytecode to change.

3. **Clear Cache**: Clear the compiler cache in Hardhat by running `npx hardhat clean`. This can help in cases where the build artifacts are not updating correctly.

4. **Recompile with Verbose Logging**: Recompile your contracts with verbose logging enabled to see more detailed information about what the compiler is doing. This might give you insights into why the bytecode is changing.

5. **Manual Bytecode Comparison**: After recompiling, manually compare the bytecode of `UpgradeableOpenfortProxy` from different compilations to pinpoint what is changing. Tools like `diffchecker` online can help with this.

6. **Review Compiler Version**: Since you are using `zksolc` version 1.5.2, ensure this version is stable and consider testing with another version if persistent issues are noted.

If after these steps the issue persists, it might be beneficial to reach out directly to the ZKsync support or community with detailed information about your setup and the steps you've taken to isolate the issue. This can help in identifying if there's a deeper issue with the toolchain or compiler.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by uF4No** [comment]
Please try using the `forceEVMLA: true in the zksolc settings of the hardhat config file

**Comment by hedgar2017** [comment, marked as answer]
@uF4No we've investigated is deeply, and it turned out that *there does exist a dependency relation* between the abovementioned contracts:
```
contracts/core/upgradeable/UpgradeableOpenfortProxy.sol -> contracts/core/upgradeable/UpgradeableOpenfortFactory.sol -> contracts/core/upgradeable/UpgradeableOpenfortAccount.sol -> contracts/core/base/BaseRecoverableAccount.sol -> contracts/core/base/BaseOpenfortAccount.sol
```
So both *solc* and *zksolc* are working fine.

**Reply by Haypierre** [reply]
Thanks a lot @hedgar2017  for taking the time to investigate
